CONTRIBUTING.md

```diff
@@ -3,16 +3,57 @@
-5. git commit your changes
-6. git push origin HEAD
-7. To request merge into master send a pull request from the web ui
-https://github.com/Microsoft/onnxruntime.
-8. Add 'Microsoft/onnxruntime' as a reviewer.
+# Proposing new public APIs
+The ONNX runtime has a collection of [public API's](docs/HighLevelDesign.md).  Some of these API's make their way back into the Windows OS.  We make compatability committments for these API's and follow a structured process when adding to them.  Please use the [Feature Request issue template](issues/new?template=feature_request.md) before starting any PR's that affect any of the public API's.
+
+# Process details
+
+Please search the [issue tracker](https://github.com/microsoft/onnxruntime/issues) for a similar idea first: there may already be an issue you can contribute to.
+
+1. **Create Issue**  
+All code changes must be tied to an Issue. 
+To propose a new feature or API please start by filing a new issue in the [issue tracker](https://github.com/microsoft/onnxruntime/issues).  
+Include as much detail as you have. It's fine if it's not a complete design: just a summary and rationale is a good starting point.
+
+2. **Wait for Team Owner**  
+We will assign an ONNX Runtime team owner to your issue. The ONNX Runtime team regularly triages all incoming issues.  
+
+3. **Discussion**  
+We'll keep the issue open for community discussion until the team owner decides it's ready or should be closed.  
+Note that if an issue isn't a high priority or has many open questions then it might stay open for a long time.
+
+4. **Owner Review**  
+The ONNX Runtime team will review the proposal and either approve or close the issue based on whether it broadly aligns with the [Onnx Runtime Roadmap](../docs/Roadmap.md) and contribution guidelines.
+
+5. **API Review**  
+If the feature adds new APIs then we'll start an API review. 
+All new public APIs must be reviewed before merging.  
+
+6. **Implementation**  
+A feature can be implemented by you, the ONNX Runtime team, or other community members.  
+Code contributions are greatly appreciated: feel free to work on any reviewed feature you proposed, or choose one in the backlog and send us a PR. Please let us know in the issue comments if you are actively working on implementing a feature so we can ensure it's assigned to you.   
+
+7. **Checkin Procedure**  
+Once a feature is complete and tested according to the contribution guidelines follow these steps:
+
+ * Fork the repo
+ * git clone your fork
+ * Create feature branch
+ * Make and checkin your changes along with unit tests
+ * git commit your changes
+ * git push origin HEAD
+ * To request merge into master, send a pull request from the [web ui](https://github.com/Microsoft/onnxruntime).
+ * Add 'Microsoft/onnxruntime' as a reviewer.
+
+8. **Documentation and sample updates**  
+We will update the documentation and if applicable add a sample to the samples repository.
```

Comment:
The person who implements the functionality should be responsible for adding corresponding documentation and samples.

Reply:
Agreed. I moved the docs/samples section prior to check-in and modified the wording to reflect your suggestion.